app/backend/src/couchers/query.py

```diff
@@ -0,0 +1,59 @@
+from sqlalchemy import or_
+from sqlalchemy.orm import Query, aliased
+
+from couchers.models import User, UserBlock
+from couchers.utils import is_valid_email, is_valid_user_id, is_valid_username
+
+
+class CouchersQuery(Query):
+    def _blocked_users(self, session, user_id):
+        """
+        Gets list of blocked user IDs or users that have blocked this user: those should be hidden
+        """
+        relevant_user_blocks = (
+            session.query(UserBlock)
+            .filter(or_(UserBlock.blocking_user_id == user_id, UserBlock.blocked_user_id == user_id))
+            .all()
+        )
+
+        return [
+            user_block.blocking_user_id if user_block.blocking_user_id != user_id else user_block.blocked_user_id
+            for user_block in relevant_user_blocks
+        ]
+
+    def filter_by_username_or_email(self, field):
+        if is_valid_username(field):
+            return self.filter(User.username == field)
+        elif is_valid_email(field):
+            return self.filter(User.email == field)
+        # no fields match, this will return no rows
+        return self.filter(False)
+
+    def filter_by_username_or_id(self, field):
+        if is_valid_username(field):
+            return self.filter(User.username == field)
+        elif is_valid_user_id(field):
+            return self.filter(User.id == field)
+        # no fields match, this will return no rows
+        return self.filter(False)
+
+    def filter_users(self, context, table=User):
+        """
+        Filters out users that should not be visible: blocked, deleted, or banned
+
+        Filters the given table, assuming it's already joined/selected from
+        """
+        hidden_users = self._blocked_users(self.session, context.user_id)
```

Comment:
This call does a full sql query to the database, it would be better if it could augment the current query with the right filters. But it could be an improvement to be done later.

Reply:
I'll leave it for later